File: leetcode/python/2641.cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root
        
        levelSums = [0]
        def dfs(node:Optional[TreeNode], depth):
            if node is None:
                return
            if depth == len(levelSums):
                levelSums.append(0)
            
            levelSums[depth] += node.val

            dfs(node.left, depth+1)
            dfs(node.right, depth+1)
        
        dfs(root, 0)
        
        def buildTree(root: Optional[TreeNode], depth:int, siblingSum:int) -> Optional[TreeNode]:
            if not root:
                return None
            ans = TreeNode(val=levelSums[depth]-siblingSum)
            childSum = (0 if root.left is None else root.left.val) + (0 if root.right is None else root.right.val)
            if root.left:
                ans.left = buildTree(root.left, depth+1, childSum)
            if root.right:
                ans.right = buildTree(root.right, depth+1, childSum)
            
            return ans
        
        return buildTree(root, 0, root.val)

File: leetcode/cpp/2641.cousins-in-binary-tree-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int levelSums[100001];
    TreeNode* replaceValueInTree(TreeNode* root) {
        // vector<int> levelSums(100000, 0);
        dfs(root, 0);
        return buildTree(root, 0, levelSums[0]);
    }
private:
    TreeNode* buildTree(TreeNode* root, int depth, int siblingSum) {
        if (root == nullptr) {
            return nullptr;
        }
        TreeNode* ans = new TreeNode(levelSums[depth] - siblingSum);
        int childSum = 0;
        if (root->left != nullptr) {
            childSum += root->left->val;
        }
        if (root->right != nullptr) {
            childSum += root->right->val;
        }
        ans->left = buildTree(root->left, depth+1, childSum);
        ans->right = buildTree(root->right, depth+1, childSum);
        return ans;
    }
    void dfs(TreeNode* root, int depth) {
        if (root == nullptr) {
            return;
        }
        levelSums[depth] += root->val;
        dfs(root->left, depth + 1);
        dfs(root->right, depth + 1);
    }
};

